### PR TITLE
Purecn/run

### LINKED ISSUE #2940: new module: purecn/run
### Is there an existing module for this?

- [X] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by aldosr**
## PR checklist

Closes #2940  

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by lbeltrame** [comment]
Blocked by https://github.com/BioContainers/multi-package-containers/pull/2554.

**Comment by lbeltrame** [comment]
Blocked by https://github.com/BioContainers/multi-package-containers/pull/2554.

File: modules/nf-core/purecn/run/main.nf [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    input:
+    tuple val(meta), path(vcf)
+    path tumor_coverage
+    path normal_coverage
+    path normal_db
+    path intervals
+    path mapping_bias
+    path blacklist
+    val genome
```
Comment (lbeltrame) [inline modules/nf-core/purecn/run/main.nf:19]: These, and the other `val` lines can be supplied as `$args`, no need to have them as parameters.
`mapping bias` is also optional if I recall, so the same thing applies, likewise for `blacklist`. This should simplify the inputs a lot.

**Review by lbeltrame — CHANGES_REQUESTED** [review]

File: modules/nf-core/purecn/run/main.nf (hunk 1 above)
Comment (aldosr) [reply modules/nf-core/purecn/run/main.nf:14]: Changes done.

**Comment by nvnieuwk** [comment]
Yeah you can use a stub run :) just make sure the module works with a bigger dataset locally every time you update the module

**Comment by aldosr** [comment]
> Yeah you can use a stub run :) just make sure the module works with a bigger dataset locally every time you update the module

I think we are fine. Anyway, due to the following [PR](https://github.com/nf-core/proteinfold/pull/106) I had to remove the "quay.io" from the container string. Should we open a PR for the previous modules?

File: modules/nf-core/purecn/run/main.nf [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+    """
+    library_path=\$(Rscript -e 'cat(.libPaths(), sep = "\\n")')
+    Rscript "\$library_path"/PureCN/extdata/PureCN.R \\
+        --out ./ \\
+        --tumor ${coverage} \\
+        --sampleid ${prefix} \\
+        --vcf ${vcf} \\
+        --normaldb ${normal_db} \\
+        --intervals ${intervals} \\
```
Comment (lbeltrame) [inline modules/nf-core/purecn/run/main.nf:61]: Check if any of these arguments are optional, and if so, they should be crammed into `$args` instead.

File: modules/nf-core/purecn/run/meta.yml [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+      pattern: "*_amplification_pvalues.csv"
+  - chr_pdf:
+      type: file
+      description: |
+        PDF file containing chromosome plots
+      pattern: "*_chromosomes.pdf"
+  - seg:
+      type: file
+      description: |
+        Segmentation file generated from DNAcopy.R
```
Comment (lbeltrame) [inline modules/nf-core/purecn/run/meta.yml:81]: I'm not sure this is actually true for all segmentation methods. Can you look at the PureCN code and see whether `_dnacopy.seg` is appended regardless of the segmentation method used?

**Review by lbeltrame — COMMENTED** [review]
Almost there IMO.

File: modules/nf-core/purecn/run/main.nf (hunk 2 above)
Comment (aldosr) [reply modules/nf-core/purecn/run/main.nf:61]: According to [Purecn best practices](https://www.bioconductor.org/packages/devel/bioc/vignettes/PureCN/inst/doc/Quick.html#43_PureCN) this is the minimal test run.

File: modules/nf-core/purecn/run/meta.yml (hunk 3 above)
Comment (aldosr) [reply modules/nf-core/purecn/run/meta.yml:81]: It is, if I'm not wrong: https://github.com/lima1/PureCN/blob/d90f3f0e80fde4b1ba70ad6e99d1870baf708157/inst/extdata/PureCN.R#L458

**Review by lbeltrame — APPROVED** [review]

File: modules/nf-core/purecn/run/meta.yml [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+      pattern: "*.pdf"
+  - local_optima_pdf:
+      type: file
+      description: |
+        PDF file containing local optima plots
+      pattern: "*_local_optima.pdf"
+  - seg:
+      type: file
+      description: |
+        Segmentation file generated from DNAcopy.R
```
Comment (lbeltrame) [inline modules/nf-core/purecn/run/meta.yml:56]: Change to "Tab-delimited file containing segmentation results."

File: modules/nf-core/purecn/run/meta.yml [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+      pattern: "*_chromosomes.pdf"
+  - segmentation_pdf:
+      type: file
+      description: |
+        PDF file containing segmentation plots
+      pattern: "*_segmentation.pdf"
+  - multisample.seg:
+      type: file
+      description: |
+        Segmentation file generated from DNAcopy.R
```
Comment (lbeltrame) [inline modules/nf-core/purecn/run/meta.yml:96]: Change to "Tab-delimited file containing segmentation data from multiple samples". DNAcopy.R is misleading given that multiple segmentation algorithms can be used here.

**Comment by aldosr** [comment]
@nvnieuwk Can you take a look please :) ?

**Review by nvnieuwk — APPROVED** [review]
LGTM! Awesome work!

**Pull request merged**